oak/server/rust/oak_runtime/src/node/wasm.rs

```diff
@@ -0,0 +1,673 @@
+use log::{debug, error};
+
+use alloc::string::String;
+use alloc::sync::Arc;
+
+use wasmi::ValueType::*;
+
+use byteorder::{ByteOrder, LittleEndian};
+
+use rand::RngCore;
+
+use oak_abi::OakStatus;
+
+use crate::channel::{ChannelEither, ChannelReader, ChannelWriter, TakeMessage};
+use crate::{HashMap, OakMessage, Runtime};
+
+// 0: node_create: (usize, usize, usize, usize, u64) -> u32
+// 1: random_get: (usize, usize) -> u32
+// 2: channel_close: (u64) -> u32
+// 3: channel_create: (usize, usize) -> u32
+// 4: channel_write: (u64, usize, usize, usize, u32) -> u32
+// 5: channel_read: (u64, usize, usize, usize, usize, u32, usize) -> u32
+// 6: wait_on_channels: (usize, u32) -> u32
+
+const NODE_CREATE: usize = 0;
+const RANDOM_GET: usize = 1;
+const CHANNEL_CLOSE: usize = 2;
+const CHANNEL_CREATE: usize = 3;
+const CHANNEL_WRITE: usize = 4;
+const CHANNEL_READ: usize = 5;
+const WAIT_ON_CHANNELS: usize = 6;
+
+/// TODO: doc
+/// TODO: randomized handle allocator
+struct WasmInterface {
+    runtime: Arc<Runtime>,
+
+    readers: HashMap<u64, ChannelReader>,
+    writers: HashMap<u64, ChannelWriter>,
+
+    next_handle: u64,
+
+    memory: Option<wasmi::MemoryRef>,
+}
+
+impl WasmInterface {
+    /// TODO: doc
+    pub fn new(runtime: &Arc<Runtime>, initial_handle: ChannelReader) -> WasmInterface {
+        WasmInterface {
+            runtime: runtime.clone(),
+            readers: {
+                let mut r = HashMap::new();
+                r.insert(1, initial_handle);
+                r
+            },
+            writers: HashMap::new(),
+            next_handle: 2,
+            memory: None,
+        }
+    }
+
+    /// TODO: doc
+    fn node_create(
+        &self,
+        name_ptr: u32,
```

Comment:
For the existing runtime we tried to be quite careful about the types on the ABI, so any future shift to Wasm64 would hopefully be trivial;  I think this needs similar thought/changes throughout, cf. 4467abb5ddee0c70b749e29dd404eb8d125a3e2c